File: examples/python/17_response_caching.py
"""
Response Caching Example

Demonstrates client-side response caching to reduce API costs
by caching identical requests. Uses qwen/qwen3-32b on OpenRouter.

Cache hit = no API call = free!

Note: We use ThinkingConfig.disabled() to disable the model's reasoning mode,
which allows using lower max_tokens values efficiently.

Requirements:
- Set OPENROUTER_API_KEY environment variable
- uv sync

Run:
    uv run python 17_response_caching.py
"""

from llmkit import LLMKitClient, Message, CompletionRequest, ThinkingConfig


def main():
    # Note: Python binding may not expose CachingProvider directly.
    # This example demonstrates the concept using simple memoization.
    # For full caching support, use the Rust or Node.js examples.

    client = LLMKitClient.from_env()
    model = "openrouter/qwen/qwen3-32b"

    # Simple in-memory cache for demonstration
    cache = {}

    def cached_complete(request_key: str, request: CompletionRequest):
        """Complete with simple caching."""
        if request_key in cache:
            print("  [CACHE HIT - no API call]")
            return cache[request_key]
        else:
            print("  [CACHE MISS - API call]")
            response = client.complete(request)
            cache[request_key] = response
            return response

    print("Response Caching Example")
    print("=" * 50)

    # First request - will hit the API
    print("\n1. First request (France):")
    print("-" * 40)

    request1 = CompletionRequest(
        model=model,
        messages=[Message.user("What is the capital of France?")],
        system="Answer briefly in one sentence.",
        max_tokens=100,
    ).with_thinking_config(ThinkingConfig.disabled())
    response = cached_complete("france", request1)
    print(f"Response: {response.text_content().strip()}")

    # Second identical request - should hit the cache
    print("\n2. Same request again (France):")
    print("-" * 40)

    request2 = CompletionRequest(
        model=model,
        messages=[Message.user("What is the capital of France?")],
        system="Answer briefly in one sentence.",
        max_tokens=100,
    ).with_thinking_config(ThinkingConfig.disabled())
    response = cached_complete("france", request2)
    print(f"Response: {response.text_content().strip()}")

    # Third request with different question - cache miss
    print("\n3. Different question (Germany):")
    print("-" * 40)

    request3 = CompletionRequest(
        model=model,
        messages=[Message.user("What is the capital of Germany?")],
        system="Answer briefly in one sentence.",
        max_tokens=100,
    ).with_thinking_config(ThinkingConfig.disabled())
    response = cached_complete("germany", request3)
    print(f"Response: {response.text_content().strip()}")

    # Fourth request - same as first (cache hit)
    print("\n4. First question again (France):")
    print("-" * 40)

    request4 = CompletionRequest(
        model=model,
        messages=[Message.user("What is the capital of France?")],
        system="Answer briefly in one sentence.",
        max_tokens=100,
    ).with_thinking_config(ThinkingConfig.disabled())
    response = cached_complete("france", request4)
    print(f"Response: {response.text_content().strip()}")

    # Summary
    print("\n" + "=" * 50)
    print("Summary:")
    print("  Total requests: 4")
    print("  API calls made: 2 (misses)")
    print("  Free responses: 2 (hits)")
    print("  Hit rate: 50%")
    print("\nCaching saves API costs on repeated identical requests!")


if __name__ == "__main__":
    main()

File: examples/17_response_caching.rs
//! Response Caching Example
//!
//! Demonstrates client-side response caching to reduce API costs
//! by caching identical requests. Uses qwen/qwen3-32b on OpenRouter.
//!
//! Cache hit = no API call = free!
//!
//! Requirements:
//! - Set OPENROUTER_API_KEY environment variable
//!
//! Run:
//!     cargo run --example 17_response_caching

use llmkit::cache::{CacheConfig, CachingProvider, InMemoryCache};
use llmkit::providers::OpenRouterProvider;
use llmkit::{CompletionRequest, LLMKitClient, Message};
use std::sync::Arc;
use std::time::Duration;

#[tokio::main]
async fn main() -> llmkit::Result<()> {
    // Create the base provider
    let provider = OpenRouterProvider::from_env()?;

    // Wrap with caching - caches identical requests for 1 hour
    let cache = InMemoryCache::new(
        CacheConfig::new()
            .with_ttl(Duration::from_secs(3600))
            .with_max_entries(1000),
    );
    let cached_provider = Arc::new(CachingProvider::new(provider, cache));

    // Create client with the cached provider
    let client = LLMKitClient::builder()
        .with_provider("openrouter", cached_provider.clone())
        .build()
        .await?;

    let model = "openrouter/qwen/qwen3-32b";

    println!("Response Caching Example");
    println!("{}", "=".repeat(50));

    // First request - will hit the API
    println!("\n1. First request (cache MISS - API call):");
    println!("{}", "-".repeat(40));

    let request =
        CompletionRequest::new(model, vec![Message::user("What is the capital of France?")])
            .with_system("Answer briefly in one sentence.")
            .without_thinking()
            .with_max_tokens(100);

    let response = client.complete(request.clone()).await?;
    let stats1 = cached_provider.stats();
    println!("Response: {}", response.text_content().trim());
    println!(
        "Cache stats: {} hits, {} misses",
        stats1.hits, stats1.misses
    );

    // Second identical request - should hit the cache
    println!("\n2. Identical request (cache HIT - no API call!):");
    println!("{}", "-".repeat(40));

    let request =
        CompletionRequest::new(model, vec![Message::user("What is the capital of France?")])
            .with_system("Answer briefly in one sentence.")
            .without_thinking()
            .with_max_tokens(100);

    let response = client.complete(request).await?;
    let stats2 = cached_provider.stats();
    println!("Response: {}", response.text_content().trim());
    println!(
        "Cache stats: {} hits, {} misses",
        stats2.hits, stats2.misses
    );

    // Third request with different question - cache miss
    println!("\n3. Different question (cache MISS - API call):");
    println!("{}", "-".repeat(40));

    let request = CompletionRequest::new(
        model,
        vec![Message::user("What is the capital of Germany?")],
    )
    .with_system("Answer briefly in one sentence.")
    .without_thinking()
    .with_max_tokens(100);

    let response = client.complete(request).await?;
    let stats3 = cached_provider.stats();
    println!("Response: {}", response.text_content().trim());
    println!(
        "Cache stats: {} hits, {} misses",
        stats3.hits, stats3.misses
    );

    // Fourth request - same as first (cache hit)
    println!("\n4. Same as first question (cache HIT):");
    println!("{}", "-".repeat(40));

    let request =
        CompletionRequest::new(model, vec![Message::user("What is the capital of France?")])
            .with_system("Answer briefly in one sentence.")
            .without_thinking()
            .with_max_tokens(100);

    let response = client.complete(request).await?;
    let stats4 = cached_provider.stats();
    println!("Response: {}", response.text_content().trim());
    println!(
        "Cache stats: {} hits, {} misses",
        stats4.hits, stats4.misses
    );

    // Summary
    println!("\n{}", "=".repeat(50));
    println!("Summary:");
    println!("  Total requests: 4");
    println!("  API calls made: {} (misses)", stats4.misses);
    println!("  Free responses: {} (hits)", stats4.hits);
    println!("  Hit rate: {:.0}%", stats4.hit_rate() * 100.0);
    println!("\nCaching saves API costs on repeated identical requests!");

    Ok(())
}
